libmisc/user_busy.c

```diff
@@ -106,6 +107,54 @@ static int user_busy_utmp (const char *name)
 #endif				/* !__linux__ */
 
 #ifdef __linux__
+#ifdef ENABLE_SUBIDS
+#define in_parentuid_range(uid) ((uid) >= parentuid && (uid) < parentuid + range)
+static int in_mapping (const char *sname, unsigned long ruid, unsigned long euid, unsigned long suid)
+{
+	/* 41: /proc/xxxxxxxxxx/task/xxxxxxxxxx/uid_map + \0 */
+	char path[41];
+	char buf[512], buf2[512];
+	ssize_t llen1, llen2;
+	FILE *file;
+
+	snprintf (path, 41, "/proc/%s/ns/user", sname);
+
+	if ((llen1 = readlink (path, buf, sizeof(buf))) == -1)
+		return 0;
+
+	if ((llen2 = readlink ("/proc/self/ns/user", buf2, sizeof(buf2))) == -1)
+		return 0;
+
+	if (llen1 == llen2 && memcmp(buf, buf2, llen1) == 0)
+		return 0; /* same namespace */
+
+	snprintf (path, 41, "/proc/%s/uid_map", sname);
+
+	file = fopen (path, "r");
+	if (NULL == file) {
+		return 0;
+	}
+	if (fgets (buf, sizeof (buf), file) == buf) {
+		unsigned long nsuid, parentuid, range;
+
+		if (sscanf (buf,
+			    " %lu %lu %lu\n",
+			    &nsuid, &parentuid, &range) == 3) {
+			if (   in_parentuid_range(ruid)
+			    || in_parentuid_range(euid)
+			    || in_parentuid_range(suid) ) {
```

Comment:
...  what is the case where you expect these not to be true?  The task would have to be running as a uid which not mapped into the task's own uid mapping?  Which is unlikely in the general container case since we've already established that one if its uids is in the victim's subuid range.  Can you explain what case this is meant to check?

Reply:
The thing is I've originally implemented just this check and wanted to exempt just the case where all the uid range is mapped (i.e. as is for the regular non-namespaced tasks). But then I realized the logic would be incorrect for the case when the namespace is the same. So I've implemented the namespace comparison.

So you're right that this check actually does not make much sense now and I'm going to remove it.